codalab/worker/dependency_manager.py

```diff
@@ -15,33 +17,76 @@
 from codalab.worker.un_tar_directory import un_tar_directory
 from codalab.worker.fsm import BaseDependencyManager, DependencyStage, StateTransitioner
 from codalab.worker.worker_thread import ThreadDict
-from codalab.worker.state_committer import JsonStateCommitter
 from codalab.worker.bundle_state import DependencyKey
+from codalab.worker.state_committer import JsonStateCommitter
+
+logging.getLogger('flufl.lock').setLevel(logging.WARNING)
+from flufl.lock import Lock, AlreadyLockedError, NotLockedError  # noqa: E402
+
 
 logger = logging.getLogger(__name__)
 
 DependencyState = namedtuple(
-    'DependencyState', 'stage dependency_key path size_bytes dependents last_used message killed'
+    'DependencyState',
+    'stage downloading_by dependency_key path size_bytes dependents last_used last_downloading message killed',
 )
 
+DependencyManagerState = Dict[str, Union[Dict[DependencyKey, DependencyState], Set[str]]]
+
 
 class DownloadAbortedException(Exception):
     """
-
     Exception raised by the download if a download is killed before it is complete
     """
 
     def __init__(self, message):
         super(DownloadAbortedException, self).__init__(message)
 
 
+class NFSLock:
+    def __init__(self, path):
+        # Specify the path to a file that will be used to synchronize the lock.
+        # Per the flufl.lock documentation, use a file that does not exist.
+        self._lock = Lock(path)
+
+        # Locks have a lifetime (default 15 seconds) which is the period of time that the process expects
+        # to keep the lock once it has been acquired. We set the lifetime to be 30 minutes as we expect
+        # all operations that require locks to be completed within that time.
+        self._lock.lifetime = timedelta(minutes=30)
+
+        # Ensure multiple threads within a process run NFSLock operations one at a time.
+        # Multiple threads accessing flufl lock can cause slow performance.
+        # We must acquire the reentrant lock before acquiring the flufl lock and only release after
+        # the flufl lock is released.
```

Comment:
@percyliang This is the change Tony and I added last Wednesday that we discussed last meeting.